### PR TITLE
[ISSUE-76] Disallow sendShuffleData if requireBufferId expired

### PR DESCRIPTION
**Pull request by xianjingfeng**
### What changes were proposed in this pull request?

1.In server side, if requireBufferId not found when send data, thrown an exception.
2.In client side, if fail to send data, require buffer again.

### Why are the changes needed?

We found shuffle server which under high load is easy encounter java.lang.OutOfMemoryError: Java heap space even we allocate more jvm heap memory and less rss.server.buffer.capacity #76 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Have already verify in our production enviroment.

### CODE REVIEW

**Comment by jerqi** [comment]
Could we add a ut for this pr?

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java [hunk 1]
```diff
@@ -109,7 +109,7 @@ public ShuffleServerGrpcClient(String host, int port, int maxRetryAttempts, bool
     blockingStub = ShuffleServerGrpc.newBlockingStub(channel);
   }
   
-  private ShuffleServerBlockingStub getBlockingStub() {
+  public ShuffleServerBlockingStub getBlockingStub() {
```
Comment (jerqi) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:112]: Could we use `default access modifier`? It seems enough to use `package-private` .
```
ShuffleServerBlockingStub getBlockingStub()
```

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java [hunk 2]
```diff
@@ -290,10 +278,22 @@ public RssSendShuffleDataResponse sendShuffleData(RssSendShuffleDataRequest requ
...
   }
 
-  private SendShuffleDataResponse doSendData(SendShuffleDataRequest rpcRequest) {
+  private SendShuffleDataResponse doSendData(String appId, int size, RssSendShuffleDataRequest request,
+                                             Map.Entry<Integer, Map<Integer, List<ShuffleBlockInfo>>> stb,
+                                             List<ShuffleData> shuffleData) {
     int retryNum = 0;
     while (retryNum < maxRetryAttempts) {
       try {
+        long requireId = requirePreAllocation(size, request.getRetryMax(), request.getRetryIntervalMax());
```
Comment (jerqi) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:287]: Could we reuse RetryUtils?

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java (hunk 1 above)
Comment (xianjingfeng) [reply internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:112]: ut can't access this method if  use default access modifier

Comment (xianjingfeng) [reply internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:112]: ut can't access this method if  use default access modifier

Comment (jerqi) [reply internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:112]: OK

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java [hunk 3]
```diff
@@ -290,10 +278,22 @@ public RssSendShuffleDataResponse sendShuffleData(RssSendShuffleDataRequest requ
     return response;
   }
 
-  private SendShuffleDataResponse doSendData(SendShuffleDataRequest rpcRequest) {
+  private SendShuffleDataResponse doSendData(String appId, int size, RssSendShuffleDataRequest request,
```
Comment (jerqi) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcClient.java:281]: Method `doXXXXX` usually have one rpc operation. Could we encapsulate the behavior in the other places?

**Review by jerqi — APPROVED** [review]
LGTM, This is an important pr, so cc@colinmjj

**Comment by colinmjj** [comment]
To avoid possible OOM in server side, LGTM, +1

**Comment by jerqi** [comment]
Merged. Thanks @xianjingfeng

**Pull request merged**